Make RecordFunction callbacks thread local and modernize interface

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* #37639 Use TensorMethods.cpp
* #37775 Memory profiling
* #37774 Move ThreadLocalDebugInfo to c10
* #36291 Make profiler thread local
* #37745 Propagate TLS callbacks with ThreadLocalState
* #37548 Move RecordFunction into ATen
* **#37491 Make RecordFunction callbacks thread local and modernize interface**

Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

Commit:
Make RecordFunction callbacks thread local

Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

[ghstack-poisoned]

ezyang commented:
I don't have enough context to review this. There's six hundred new lines and only a single sentence describing what's in the PR. Is this related to the discussion we had about dispatcher and observers yesterday?

ilia-cher commented:
> I don't have enough context to review this. There's six hundred new lines and only a single sentence describing what's in the PR. Is this related to the discussion we had about dispatcher and observers yesterday?

it is a prerequisite, let me add more into the description

ilia-cher commented:
@ezyang with the coming changes to record function (moving it into the Dispatcher) it might be better to work not on the old implementation and try to plug it in, but with the one that are actually going to be use - the one with the new interface and support to global/thread local callbacks, and the one that I started to work on in https://github.com/pytorch/pytorch/pull/36291 - I'm splitting this PR and moving changes to the record function in this PR; 
the dispatcher integration will follow up in the next PR

ilia-cher commented:
@ezyang added more into the description

re. how this is related, for example `setShouldRun` is the functionality that we discussed yesterday - for callbacks that might want to do something other than just flipping a coin

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/autograd/record_function.h:
```diff
@@ -262,52 +347,100 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
-    double sampling_prob = 1.0,
-    std::unordered_set<RecordScope, std::hash<RecordScope>> scopes =
-        std::unordered_set<RecordScope, std::hash<RecordScope>>());
+TORCH_API uint64_t addGlobalCallback(RecordFunctionCallback cb);
```
maybe be a bit typesafe and return a wrapper struct for the handle (like c10 registrations do)

ilia-cher replied:
maybe I could go with typedef first,
If we need to generalize, I can then replace typedef with actual struct

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -1,7 +1,8 @@
+
```
remove empty line

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -11,177 +12,192 @@ namespace profiler {
-  std::uniform_real_distribution<float> dist(0.0, 1.0);
-  return dist(*gen);
+// Used to generate unique callback handles
+uint64_t next_unique_callback_id() {
```
nit: can it be just an atomic<int>? :)

ilia-cher replied:
there were some concerns voiced re. usage of atomics, but in the end record_function_benchmark will show

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -11,177 +12,192 @@ namespace profiler {
+        ++idx_c;
+        ++idx_h;
+      }
+      while ((idx_c < sorted_callbacks.size()) &&
```
can it be just "if"? then you don't need size check here.

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -11,177 +12,192 @@ namespace profiler {
+    size_t idx_c = 0;
+    size_t idx_h = 0;
+    size_t num_executed = 0;
+    while (idx_c < sorted_callbacks.size() && idx_h < sorted_handles.size()) {
```
nit: given that we usually expect all of the sorted_callbacks to be present, it could have been written as a for loop on sorted_handles that just bumps another index. But it's up to you

dzhulgakov commented on torch/csrc/autograd/record_function.h:
```diff
@@ -262,52 +347,100 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
+ * isRecordFunctionEnabled returns whether RecordFunction
+ * is enabled thread locally
+ */
+TORCH_API bool isRecordFunctionEnabled();
```
is it a temporary solution until we make recordfunction not depend on the key? If so - I'd add a comment on what will be changed in the future

ilia-cher replied:
no, this is to stay - we do want to disable record functions sometimes, e.g. within record function callbacks themselves

when we move away from TLS dispatcher key, I add another TLS flag (but it's true by default)

dzhulgakov commented on torch/csrc/autograd/record_function.h:
```diff
@@ -83,6 +85,8 @@ struct TORCH_API StringView {
 // Soft limit on the number of callbacks to use;
 constexpr std::size_t kSoftLimitCallbacks = 32;
 
+typedef c10::SmallVector<uint64_t, kSoftLimitCallbacks> handles_vector;
```
this is actually 256 bytes now - make the soft limit smaller? (usually we'd likely have <5 callbacks, so maybe 8?)

dzhulgakov reviewed (commented):
Looks good, minor comments.

Have you done sanity-check benchmark of overheads?

P.S. the merge sort algorithm is a bit magical but it's indeed a creative way to avoid refcounting callbacks

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -11,177 +11,189 @@ namespace profiler {
+      if (it != cbs.end()) {
+        // keeps it sorted
+        cbs.erase(it);
+        found = true;
```
nit: might be a bit more readable if you were just returning a bool instead of modifying `found`

dzhulgakov commented on torch/csrc/autograd/record_function.h:
```diff
@@ -81,7 +83,9 @@ struct TORCH_API StringView {
 
 // Soft limit on the number of callbacks to use;
-constexpr std::size_t kSoftLimitCallbacks = 32;
+constexpr std::size_t kSoftLimitCallbacks = 8;
```
can be even 4 (as it includes only active callbacks)

dzhulgakov commented on torch/csrc/autograd/record_function.h:
```diff
@@ -169,79 +169,164 @@ struct TORCH_API RecordFunction {
+  double sampling_prob_ = 1.0;
+  std::array<bool, static_cast<size_t>(RecordScope::NUM_SCOPES)> scopes_ = {};
+
+  static double sample_zero_one() {
```
move it to .cpp - static local variables cause havoc with dynamic linking

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

ilia-cher commented:
added record_function_benchmark results

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

Commit:
Update on "Make RecordFunction callbacks thread local and modernize interface"


Summary:
This PR modernizes RecordFunction API and adds thread local callbacks
in addition to the global ones

Changes: 
 - support for TLS callbacks, this is going to be the foundation of profiler and other tools
 - modernize interface around simple set of functions (add|remove|has|clear)(Global|ThreadLocal)(Callback) and adding RecordFunctionCallback to easily construct callbacks to be passed
 - we also add `.setShouldRun` into the callback interface to support cases when simple uniform sampling is not enough
 - to properly support add/remove introduce the idea of callback handle returned by add
 - internal implementation still uses SmallVector to store intermediate state (as before) - in this case these are vector of handles of callbacks that were picked to run
 - to speed up runtime we keep these vectors sorted, this way we can quickly enumerate callbacks that need to be run
 - added tests for new functionality

Test Plan:
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=0 USE_CUDA=0 python setup.py
develop install
./build/bin/test_jit
CI

record_function_benchmark: https://gist.github.com/ilia-cher/f1e094dae47fe23e55e7672ac4dcda2f

Differential Revision: [D21300448](https://our.internmc.facebook.com/intern/diff/D21300448)

[ghstack-poisoned]

ilia-cher commented:
(pytorch_linux_backward_compatibility_check_test was failing on trunk)
